src/client.ts

```diff
@@ -525,109 +516,88 @@ export class Client {
       const fwConstants = getFwVersionConst(this.fwVersion);
+      if (!fwConstants.maxDecoderBufSz) {
+        return cb('Please update Lattice firmware.');
+      }
       if (n < 1) {
         return cb('Must request at least one record (n=0)');
       } else if (n < 0 || startIdx < 0) {
         return cb('Both `n` and `startIdx` must be >=0');
-      } else if (!fwConstants.abiCategorySz) {
-        return cb('Outdated Lattice firmware. Please update.');
-      } else if (category.length >= fwConstants.abiCategorySz) {
-        return cb(`category must be <${fwConstants.abiCategorySz - 1} characters, got ${category.length}`);
       }
-      if (!fetched) {
-        fetched = {
-          startIdx,
-          numRemaining: null,
-          numFetched: null,
-          records: [],
-        }
-      }
-      const payload = Buffer.alloc(4 + fwConstants.abiCategorySz);
-      payload.writeUInt16LE(startIdx, 0);
-      payload.writeUInt16LE(n, 2);
-      Buffer.from(category).copy(payload, 4);
-      return this._request(payload, 'GET_ABI_RECORDS', (err, res) => {
+      const payload = Buffer.alloc(10);
+      payload.writeUInt8(decoderType, 0);
+      payload.writeUInt8(skipTotal ? 1 : 0, 1);
+      payload.writeUInt32LE(startIdx, 2);
+      payload.writeUInt32LE(n, 6);
+      return this._request(payload, 'GET_DECODERS', (err, res) => {
         if (err) {
           return cb(err);
         }
         // Correct wallet and no errors -- handle the response
-        const d = this._handleEncResponse(res, decResLengths.getAbiRecords);
-        if (d.err)
+        const d = this._handleEncResponse(res, decResLengths.getDecoders);
+        if (d.err) {
           return cb(d.err);
+        }
         // Decode the response
+        const decoders = [];
         let off = 65; // Skip 65 byte pubkey prefix
-        const numRemaining = d.data.readUInt32LE(off); off += 4;
-        const numReturned = d.data.readUInt8(off); off += 1;
-        // Start adding data if there is data to add
-        fetched.numRemaining = numRemaining;
-        if (!fetched.records) {
-          fetched.records = [];
-        }
-        for (let i = 0; i < numReturned; i++) {
-          // Parse and add the def
-          const packedDef = d.data.slice(off, off + ABI_DEF_SZ); off += ABI_DEF_SZ;
-          fetched.records.push(unpackAbiDef(packedDef));
-        }
-        // Decrement the total counter
-        opts.n -= numReturned;
-        if (opts.n < 1 || numReturned < 1) {
-          fetched.numFetched = fetched.records.length;
-          return cb(null, fetched);
-        } else {
-          // Recurse if there is more to fetch
-          opts.startIdx += numReturned;
-          return this.getAbiRecords(opts, cb, fetched);
+        const numFetched = d.data.readUInt32LE(off); off += 4;
+        const total = d.data.readUInt32LE(off); off += 4;
+        for (let i = 0; i < numFetched; i++) {
+          const sz = d.data.readUInt32LE(off); off += 4;
```

Comment:
Do we want the `off += 4` assertions on the same line?

Reply:
[756bc00](https://github.com/GridPlus/gridplus-sdk/pull/353/commits/756bc001e62e4ba9fce75734645727253137ddc2)